### PR TITLE
[rfc][sui framework] u64 module for better math support

### PR DESCRIPTION
**Pull request by sblackshear**
Trying out a new style of math support. Want to align on the API's we should provide for each integer type first, then we can create similar modules for the other integer types.

### CODE REVIEW

File: crates/sui-framework/deps/move-stdlib/sources/string.move [hunk 1]
```diff
@@ -3,8 +3,9 @@
 
 /// The `string` module defines the `String` type which represents UTF8 encoded strings.
 module std::string {
-    use std::vector;
+    use std::ascii;
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/string.move:6]: I'm guessing the changes in this file are spurious / a hold over from the other PR?

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+/// Basic math for nicer programmability
+module std::u64 {
+    use std::ascii;
+    use std::option::{Self, Option};
+    use std::string;
+    use std::vector;
+
+    /// Largest possible u64 value
+    const MAX: u64 = 0;
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:12]: I hope not!

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    public fun min(x: u64, y: u64): u64 {
+        if (x < y) {
+            x
+        } else {
+            y
+        }
+    }
+
+    /// Return the value of a base raised to a power.
+    /// This function is not overflow-safe and will abort with an overflow error if `base^exponent` > `MAX`
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:36]: This seems worth writing a `spec` for -- can we annotate this comment some way to make such things easy to find? cc @awelc

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                res = res * base;
+                exponent = exponent - 1;
+            }
+        };
+
+        res
+    }
+
+    /// Return the mean of x and y, with the remainder truncated.
+    /// This function is overflow-safe and cannot abort
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:53]: Ditto on specs here.

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    public fun mean(x: u64, y: u64): u64 {
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:54]: If there was a `min_max` function, you could define `min`, `max` and `mean` in terms of it:

```move
public fun min_max(x: u64, y: u64): (u64, u64) {
  if (x < y) {
    (x, y)
  } else {
    (y, x)
  }
}

public fun min(x: u64, y: u64): u64 {
  let (m, _) = min_max(x, y);
  m
}

public fun max(x: u64, y: u64): u64 {
  let (_, m) = min_max(x, y);
  m
}

public fun mean(x: u64, y: u64): u64 {
  let (lo, hi) = min_max(x, y);
  lo + (hi - lo) / 2
}
```

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    /// Like the + operator, but allows overflow rather instead of aborting
+    public fun saturating_add(x: u64, y: u64): u64 {
+        abort(0) // implement me
+    }
+
+    /// Compute the sum of all the values in `nums`
+    /// This function is not overflow-safe and will abort with an overflow error if the sum > `MAX`
+    public fun vec_sum(nums: &vector<u64>): u64 {
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:121]: Not sure about having these operations in here -- curious to hear other people's thoughts -- I think I'd prefer us investing in easier to use looping primitives for Move.

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        let len = vector::length(nums);
+        let total = 0;
+        while (i < len) {
+            total = total + *vector::borrow(nums, i);
+            i + i + 1;
+        };
+        total
+    }
+
+    /// Compute the mean of all the values in `nums`.
```
Comment (amnn) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:132]: nit: should document the truncation behaviour here as well.

File: crates/sui-framework/deps/move-stdlib/sources/u64.move (hunk 6 above)
Comment (sblackshear) [reply crates/sui-framework/deps/move-stdlib/sources/u64.move:121]: I think putting `vec_sum`, `vec_mean`, etc elsewhere makes sense. Curious for where folks thing `vec_sort` should live, though (here? A separate `sort` module that exposes sorting for all ground types)?

Comment (tnowacki) [reply crates/sui-framework/deps/move-stdlib/sources/u64.move:121]: Too much to all put in `vector`? 

The only convenient thing about having them here in `u64` is you can keep the name `vec_sum`. Maybe it's just reasonable just having the name `sum`? `u64::sum`

Comment (tnowacki) [reply crates/sui-framework/deps/move-stdlib/sources/u64.move:121]: My thoughts is you will probably want a sum/mean for all the integer types. And keeping them in their own module feels reasonable?

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+    /// Largest possible u64 value
+    const MAX: u64 = 0;
+
+    /// Attempting to perform an invalid operation (e.g., mean) on an empty vector
+    const EVectorEmpty: u64 = 0;
+
+    /// Return the larger of `x` and `y`
+    public fun max(x: u64, y: u64): u64 {
+        if (x > y) {
```
Comment (tnowacki) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:19]: Alternatives for all these `if`s, though I know opinions vary here
```
if (x < y) x else y
```
or 
```
if (x < y) x 
else y 
```

Comment (tnowacki) [reply crates/sui-framework/deps/move-stdlib/sources/u64.move:19]: The second is nice for the math ones below, e.g. 
```
if (x < y) x + (y - x) / 2
else y + (x - y) / 2
```

File: crates/sui-framework/deps/move-stdlib/sources/u64.move [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    /// math::sqrt(8) => 2;
+    /// math::sqrt(8 * 10000) => 282;
+    /// // now we can use this value as if it was 2.82;
+    /// // but to get the actual result, this value needs
+    /// // to be divided by 100 (because sqrt(10000)).
+    ///
+    ///
+    /// math::sqrt(8 * 1000000) => 2828; // same as above, 2828 / 1000 (2.828)
+    /// ```
+    public fun sqrt(x: u64): u64 {
```
Comment (kchalkias) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:87]: is this in practice `sqrt_ceil`?

Comment (amnn) [reply crates/sui-framework/deps/move-stdlib/sources/u64.move:87]: The docs suggest to me it's `sqrt_floor`, i.e. $sqrt(x: \mathbb{N}) = \max\\,\\{\\,y \in \mathbb{N} : y^2 \leq x \\,\\}$

Comment (kchalkias) [reply crates/sui-framework/deps/move-stdlib/sources/u64.move:87]: Yes floor, how about renaming the function to reflect that?

Comment (kchalkias) [inline crates/sui-framework/deps/move-stdlib/sources/u64.move:87]: @sblackshear I was intrigued, so I let #chatGPT to contribute to this review and compare against your scheme :)
Screen Shot 2022-12-15 at 9 27 30 PM

**Review by kchalkias — COMMENTED** [review]
From my friend chatGPT

**Comment by kklas** [comment]
`muldiv` would also be nice to have. I use it a lot in my modules (and copy paste it every time).

```
/// Calculates (a * b) / c. Aborts if the result doesn't fit into u64.
fun muldiv(a: u64, b: u64, c: u64): u64 {
    (((a as u128) * (b as u128)) / (c as u128) as u64)
}
```

It preserves precision when you're doing `(a * b) / c` where `a * b` doesn't fit into `u64` but `(a * b) / c` does, which I found to be quite often (notice it's applicable any time you have `(a / b) * c` or `a * (b / c)`). See [this](https://github.com/kunalabs-io/sui-smart-contracts/blob/master/amm/move/sources/amm.move#L173-L196) for some real life usages.

Btw there's a really cool way to calculate u256 muldiv by just using u256 registers with math tricks: https://2π.com/21/muldiv/

**Comment by damirka** [comment]
Closing as stale. Raised some of the questions from here internally.

**Pull request closed without merging**